frontend/ui/widget/menu.lua

```diff
@@ -1176,7 +1185,71 @@ function Menu:getPageNumber(item_number)
     if #self.item_table == 0 or item_number == 0 then
         return 1
     end
-    return math.ceil(math.min(item_number, #self.item_table) / self.perpage)
+    if self.items_max_lines then
+        for page, items in ipairs(self.page_items) do
+            if item_number <= items[#items] then
+                return page
+            end
+        end
+        return #self.page_items
+    else
+        return math.ceil(math.min(item_number, #self.item_table) / self.perpage)
+    end
+end
+
+function Menu:setupItemHeights()
+    if #self.item_table == 0 then
+        self.page_items = {{}}
+        return
+    end
+
+    local face = Font:getFace("smallinfofont", self.font_size)
+    local line_height = TextBoxWidget:new{
+        text = "A",
+        face = face,
+    }:getSize().h
+    local infont_size = self.items_mandatory_font_size or (self.font_size - 4)
+    local infont_face = Font:getFace("infont", infont_size)
+    local infont_char_width = TextWidget:new{
+        text = "0",
+        face = infont_face,
+        bold = true,
+    }:getSize().w
+    local available_width = self.inner_dimen.w
+    if self.is_enable_shortcut then
+        available_width = available_width - line_height - Size.span.horizontal_default
+    end
+
+    self.page_items = {} -- items in all pages
+    local items = {} -- items in a page
+    local items_height = 0 -- of all items in a page
+    for i = 1, #self.item_table do
+        local item = self.item_table[i]
+        -- exact item height can be calculated by building the TextBoxWidget for item text,
+        -- but it is slow, so estimate the number of lines by building the TextWidget
+        local item_text_width = TextWidget:new{
+            text = item.text,
+            face = face,
+            bold = item.bold,
+        }:getSize().w
+        local item_available_width = available_width - infont_char_width * (item.mandatory and #item.mandatory or 0)
+        local lines_nb = math.min(math.ceil(item_text_width / item_available_width), self.items_max_lines)
```

Comment:
I guess this could also go wrong, as TextBoxWidget will wrap on real wrap opportunities (unlike your division, which wrap anywhere), so a TextBoxWidget may need more lines than what you will compute.
If max lines is 3, you may get some items where you found that 2 lines were enough - but when rendered by TextBoxWidget, it would need 3 lines - so you will show only 2 and the truncation ellipsis :/

Reply:
Theoretically yes, practically I haven't met this in the test book.
We cannot build the correct TextBoxWidget, it is too slow, as we discussed.